src/java/com/puppetlabs/puppetserver/MetricsPuppetProfiler.java

```diff
@@ -54,29 +62,107 @@ public Set<String> getAllMetricIds() {
     }
 
     public Map<String, Timer> getFunctionTimers() {
-        return getTimers(FUNCTION_PATTERN);
+        return this.function_timers;
     }
 
     public Map<String, Timer> getResourceTimers() {
-        return getTimers(RESOURCE_PATTERN);
+        return this.resource_timers;
     }
 
     public Map<String, Timer> getCatalogTimers() {
-        return getTimers(CATALOG_PATTERN);
+        return this.catalog_timers;
     }
 
     public Map<String, Timer> getInliningTimers() {
-        return getTimers(INLINING_PATTERN);
+        return this.inlining_timers;
     }
 
     public Map<String, Timer> getPuppetDBTimers() {
-        return getTimers(PUPPETDB_PATTERN);
+        return this.puppetdb_timers;
     }
 
     @Override
     public void shutdown() {
     }
 
+    private List<String> sliceOfArrayToList(String[] idSegments, int lengthOfID) {
+        // Callers expect a mutable List returned, but Arrays.asList() returns a
+	// fix length array, which is why we have to create a List and then add to it.
+        List<String> idList = new ArrayList<String>();
+        idList.addAll(Arrays.asList(Arrays.copyOf(idSegments, lengthOfID)));
+
+	return idList;
+    }
+
+    private String safeGet(String[] collection, int i) {
+        try {
+            return collection[i];
+	}  catch (IndexOutOfBoundsException _ex) {
+            return "";
+	}
+    }
+
+    private void updateMetricsTrackers(String[] metricId, Map<String, Timer> metricsByID) {
+	String firstElement = safeGet(metricId, 0);
+	String secondElement = safeGet(metricId, 1);
+
+        if ("functions".equals(firstElement)) {
+            Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 2)));
+            this.function_timers.put(secondElement, metric);
+
+        } else if ("compiler".equals(firstElement)) {
+            String thirdElemet = safeGet(metricId, 2);
+
+            if ("evaluate_resource".equals(secondElement)) {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 3)));
+                this.resource_timers.put(thirdElemet, metric);
+
+            } else if ("static_compile_inlining".equals(secondElement)) {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 3)));
+                this.inlining_timers.put(thirdElemet, metric);
+
+            } else {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 2)));
+                this.catalog_timers.put(secondElement, metric);
+            }
+
+        } else if ("puppetdb".equals(firstElement)) {
+            if ("query".equals(secondElement)) {
+                Timer metric = metricsByID.get(getMetricName(sliceOfArrayToList(metricId, 2)));
+                this.puppetdb_timers.put(secondElement, metric);
+
+	    } else {
+		String thirdElemet = safeGet(metricId, 2);
+
+                if (
```

Comment:
Its probably overkill, but i wonder if it would be worth comparing metrics from two reference workflows. Matching the regex to this logic is hurting my brain. It may not matter too much though... These are hard to reason about 🫠

Reply:
The regex matching definitely hurt my brain as well! I got it wrong a couple of times. I can run the acceptance test and collect what metrics are reported before and after this change. I've ran it after to make sure they look right (see this comment https://github.com/puppetlabs/puppetserver/pull/2880#discussion_r1778993676) , but haven't done a full comparison. I'll do that.

Reply:
Here are the previous metrics from 7.17.2:
```
{"service_version":"7.17.2","service_status_version":1,"detail_level":"debug","state":"running","status":{"experimental":{"function-metrics":[{"function":"puppetdb_query","count":1,"mean":369,"aggregate":369},{"function":"lookup","count":2,"mean":45,"aggregate":90}],"resource-metrics":[{"resource":"Class[main]","count":2,"mean":193,"aggregate":386}],"catalog-metrics":[{"metric":"compile","count":2,"mean":886,"aggregate":1772},{"metric":"find_node","count":2,"mean":11,"aggregate":22}],"puppetdb-metrics":[{"metric":"query","count":1,"mean":366,"aggregate":366},{"metric":"resource_search","count":8,"mean":31,"aggregate":248},{"metric":"command_submit_replace_catalog","count":2,"mean":68,"aggregate":136},{"metric":"command_submit_store_report","count":2,"mean":51,"aggregate":102},{"metric":"command_submit_replace_facts","count":2,"mean":43,"aggregate":86},{"metric":"catalog_munge","count":2,"mean":9,"aggregate":18},{"metric":"report_convert_to_wire_format_hash","count":2,"mean":5,"aggregate":10},{"metric":"facts_encode","count":2,"mean":2,"aggregate":4}],"inline-metrics":[]}},"active_alerts":[],"service_name":"puppet-profiler"}
```
and here are the metrics returned after this pr:
```
{"service_version":"7.17.3-SNAPSHOT","service_status_version":1,"detail_level":"debug","state":"running","status":{"experimental":{"function-metrics":[{"function":"puppetdb_query","count":1,"mean":411,"aggregate":411},{"function":"lookup","count":2,"mean":46,"aggregate":92}],"resource-metrics":[{"resource":"Node[weakest-byword.delivery.puppetlabs.net]","count":1,"mean":421,"aggregate":421},{"resource":"Class[main]","count":2,"mean":184,"aggregate":368},{"resource":"Node[resource-exporter.test]","count":1,"mean":208,"aggregate":208}],"catalog-metrics":[{"metric":"compile","count":2,"mean":912,"aggregate":1824},{"metric":"evaluate_ast_node","count":2,"mean":321,"aggregate":642},{"metric":"evaluate_main","count":2,"mean":188,"aggregate":376},{"metric":"evaluate_generators","count":2,"mean":115,"aggregate":230},{"metric":"iterate_on_generators","count":3,"mean":75,"aggregate":225},{"metric":"evaluate_collections","count":2,"mean":108,"aggregate":216},{"metric":"create_settings_scope","count":2,"mean":54,"aggregate":108},{"metric":"set_node_params","count":2,"mean":14,"aggregate":28},{"metric":"find_facts","count":2,"mean":7,"aggregate":14},{"metric":"find_node","count":2,"mean":6,"aggregate":12},{"metric":"finish_catalog","count":2,"mean":1,"aggregate":2},{"metric":"init_server_facts","count":1,"mean":0,"aggregate":0},{"metric":"validate_pre_finish","count":2,"mean":0,"aggregate":0},{"metric":"evaluate_node_classes","count":2,"mean":0,"aggregate":0},{"metric":"validate_final","count":2,"mean":0,"aggregate":0},{"metric":"evaluate_definitions","count":3,"mean":0,"aggregate":0}],"puppetdb-metrics":[{"metric":"query","count":1,"mean":409,"aggregate":409},{"metric":"resource_search","count":8,"mean":36,"aggregate":288},{"metric":"facts_find","count":2,"mean":139,"aggregate":278},{"metric":"catalog_save","count":2,"mean":131,"aggregate":262},{"metric":"facts_save","count":2,"mean":91,"aggregate":182},{"metric":"command_submit_replace_catalog","count":2,"mean":67,"aggregate":134},{"metric":"command_submit_replace_facts","count":2,"mean":56,"aggregate":112},{"metric":"report_process","count":2,"mean":50,"aggregate":100},{"metric":"command_submit_store_report","count":2,"mean":39,"aggregate":78},{"metric":"payload_format","count":8,"mean":3,"aggregate":24},{"metric":"facts_encode","count":2,"mean":10,"aggregate":20},{"metric":"catalog_munge","count":2,"mean":10,"aggregate":20},{"metric":"report_convert_to_wire_format_hash","count":2,"mean":4,"aggregate":8}],"inline-metrics":[]}},"active_alerts":[],"service_name":"puppet-profiler"}
```

These aren't pretty printed, not as an f-u, but because I think putting them pretty printed in a github comment would make the content very hard to compare. I put them into files, piped them through jq, and compared two terminals side by side. I think they look correct.